addons/web_editor/static/src/js/wysiwyg/MoveNodePlugin.js

```diff
@@ -0,0 +1,374 @@
+
+import SmoothScrollOnDrag from 'web/static/src/js/core/smooth_scroll_on_drag.js';
+import {
+    ancestors,
+    setSelection,
+    closestElement,
+    resetUnbreakable,
+} from '@web_editor/js/editor/odoo-editor/src/OdooEditor';
+
+const widgetContainerWidth = 25;
+const widgetMoveSize = 20;
+
+const allowedElements = 'h1, h2, h3, p, hr, pre, blockquote, ul, ol, table, .o_knowledge_behavior_anchor, .o_text_columns';
+function isNodeMovable(node) {
+    return node.parentElement &&
+        node.parentElement.getAttribute('contentEditable') === 'true';
+}
+
+function isPointInside(rect, x, y) {
+    return rect.left <= x &&
+        rect.right >= x &&
+        rect.top <= y &&
+        rect.bottom >= y;
+};
+
+export class MoveNodePlugin {
+    constructor(options = {}) {
+        this._options = options;
+    }
+
+    start() {
+        this._editor = this._options.editor;
+        this._editable = this._options.editor.editable;
+        this._document = this._options.editor.document;
+
+        this._editor.addDomListener(this._editable, 'mousemove', this._onMousemove.bind(this), true);
+        this._editor.addDomListener(this._editor.document, 'keydown', this._onDocumentKeydown.bind(this), true);
+        this._editor.addDomListener(this._editor.document, 'mousemove', this._onDocumentMousemove.bind(this), true);
+
+        const positionOption = {
+            position: 'before',
+            targetContainerId: 'oe-avatars-overlap-container'
+        };
+        // This container help to add zone into which the mouse can activate the move widget.
+        this._widgetHookContainer = this._editor.makeAbsoluteContainer('oe-widget-hooks-container', positionOption);
+        // This container contains the differents widgets.
+        this._widgetContainer = this._editor.makeAbsoluteContainer('oe-widgets-container', positionOption);
+        // This container contains the jquery helper element.
+        this._dragHelperContainer = this._editor.makeAbsoluteContainer('oe-movenode-helper-container', positionOption);
+        // This container contains drop zones. They are the zones that handle where the drop should happen.
+        this._dropzonesContainer = this._editor.makeAbsoluteContainer('oe-dropzones-container', positionOption);
+        // This container contains drop hint. The final rectangle showed to the user.
+        this._dropzoneHintContainer = this._editor.makeAbsoluteContainer('oe-dropzone-hint-container', positionOption);
+
+        // Uncomment line for debugging tranparent zones
+        // this._widgetHookContainer.classList.add('debug');
+        // this._dropzonesContainer.classList.add('debug');
+
+        this._scrollableElement = this._editable.parentElement;
+        while (this._scrollableElement) {
+            if (getComputedStyle(this._scrollableElement).overflowY === 'auto') {
+                break;
+            }
+            this._scrollableElement = this._scrollableElement.parentElement;
+        }
```

Comment:
It looks like this is a pattern that we use a lot. Maybe we need some kind of `ancestor` util ?
```js
function ancestor(node, predicate) {
    do {
        node = node?.parentElement();
    } while (node && predicate && !predicate(ancestor))
    return node;
}
```

... but then you'd need to have a look around in the code to find other places where we use this pattern to use the helper function instead.

Reply:
Problem with our closestElement util: it limit the node to be in the `odoo-editor-editable`. In this case, the node is outside the editable.

Reply:
updated to use `closestElement`